### ISSUE TITLE
Kodiak does not update a PR

### ISSUE THREAD

**Issue by staabm** [chdsbd/kodiak#261]
From time to time we experience that kodiak does not update a PR even, if new commits are made in the target branch

E.g. https://github.com/redaxo/redaxo/pull/3210#issuecomment-578396073

Our config
```
# .kodiak.toml
version = 1

[merge]
method = "squash"
delete_branch_on_merge = true
dont_wait_on_status_checks = ["WIP"] # handle github.com/apps/wip
# label to use to enable Kodiak to merge a PR
automerge_label = "automerge" # default: "automerge"
# require that the automerge label be set for Kodiak to merge a PR. if you
# disable this Kodiak will immediately attempt to merge every PR you create
require_automerge_label = true

[merge.message]
title = "pull_request_title"
body = "empty"
include_pr_number = true
strip_html_comments = true # remove html comments to auto remove PR templates

**Comment by chdsbd** [comment]
Thanks for the link, I've checked Sentry and it seems like there are some problems in parsing the webhook payloads for the repository. I've usually ignored the parsing errors but it seems like I was wrong to do so.

I think what happens for your case is that a GitHub webhook event is sent to Kodiak, which Kodiak fails to parse because Kodiak's event schema is incorrect, so Kodiak doesn't evaluate the PR.

I left a review comment https://github.com/redaxo/redaxo/pull/3210#pullrequestreview-348336182 on that linked PR and Kodiak started updating the PR, so I think this is more confirmation that the problem is in the webhook event parsing. I think Kodiak can parse the Pull Request Review webhook event, but fails at some other ones, that's why Kodiak wasn't acting on that PR.

I'm going to simplify the event parsing schemas which I believe will resolve this issue.

I also suspect that this problem may be the cause of https://github.com/chdsbd/kodiak/issues/189.

Thanks for raising this issue.

**Comment by chdsbd** [comment]
I've released #262 which should resolve this issue.

Let me know if you run into any more trouble.

**Comment by clxmstaab** [comment]
we are still seeing some PRs not getting updated.

in a private repo, the PR https://github.com/complex-gmbh/kunzmann/pull/656

![image](https://user-images.githubusercontent.com/47448731/74349702-e6366d00-4db4-11ea-9cc4-6fe345fad962.png)


as you can see on the enabled "Update Branch" there are changes in the target-branch which are not yet merged.. but kodiak does not merge them

**Comment by chdsbd** [comment]
@clxmstaab can you post your configuration file?

**Comment by chdsbd** [comment]
@clxmstaab If you change the PR (modifying the title, changing labels), does the PR get updated? I'm wondering if this is the same issue where certain events are being handled fully.

**Comment by clxmstaab** [comment]
```
# .github/.kodiak.toml
version = 1

[merge]
method = "squash"
# once a PR is merged into master, delete the branch
delete_branch_on_merge = true
# label to use to enable Kodiak to merge a PR
automerge_label = "automerge" # default: "automerge"
# require that the automerge label be set for Kodiak to merge a PR. if you
# disable this Kodiak will immediately attempt to merge every PR you create
require_automerge_label = true
# if a PR is passing all checks and is able to be merged, merge it without
# placing it in the queue. This will introduce some unfairness where those
# waiting in the queue the longest will not be served first.
prioritize_ready_to_merge = true

[merge.message]
title = "pull_request_title"
body = "github_default"
include_pr_number = true
strip_html_comments = true # remove html comments to auto remove PR templates

[update]
# update PR whenever the PR is out of date with the base branch. PR will be
# updated regardless of failing status checks, missing reviews, merge blacklist
# labels, etc. Kodiak will only update the PR if the automerge label is attached
# or `update.require_automerge_label` is false.
always = true
require_automerge_label = true
```

> If you change the PR (modifying the title, changing labels), does the PR get updated?

jep, changing labels did trigger the bot

**Comment by chdsbd** [comment]
@clxmstaab Thank you for that last bit of info!

**Comment by chdsbd** [comment]
I haven't been able to reproduce this and I'm not sure what is happening. I'm going to try to improve the logging to better associate events with installations.

Let me know if you're able to reproduce this.

**Comment by clxmstaab** [comment]
> Let me know if you're able to reproduce this.

anything special I need to focus on?

**Comment by chdsbd** [comment]
@clxmstaab if you’re able to reliably reproduce this issue or find it occurring regularly, that would be great to know. If it’s not too much trouble please update this thread when you see this happening again.

**Comment by dpraul** [comment]
I just installed kodiak for the first time in [our private repo](https://github.com/fintechstudios/website) and am seeing this as well. Config:
```toml
# .kodiak.toml
# docs: https://kodiakhq.com/docs/config-reference
version = 1

[merge]
method = "squash"

[merge.message]
title = "pull_request_title"
body = "pull_request_body"
```

![image](https://user-images.githubusercontent.com/931325/76869577-ece65300-683e-11ea-9a59-691295787660.png)

**Comment by chdsbd** [comment]
@dpraul Thanks for the report. I can't look into this in detail until later, but as a workaround you can generally edit the PR (change the title, add/remove a label) and Kodiak will receive a new webhook event from GitHub and will reexamine the PR.

**Comment by dpraul** [comment]
Yep I observed that as a workaround, just wanted to throw in another example of this happening. It has happened on two separate PRs at this point, although one of them seems to have woken up as of about 10 minutes ago.

I'll try to pay close attention to what behavior tends to trigger this.

**Comment by chdsbd** [comment]
I was testing a new release of Kodiak and I encountered this issue. Looking at the logs it seems like GitHub was reporting the [`mergeStateStatus`](https://developer.github.com/v4/enum/mergestatestatus/) as `UNKNOWN`. So I think we probably need to have Kodiak retry after a short period of time to allow GitHub to calculate mergeability.

**Comment by chdsbd** [comment]
I've released #380 which should hopefully help with this issue. I'll reopen if I get more reports.

**Comment by clxmstaab** [comment]
just realized that we hit this problem again

the (private) PR https://github.com/complex-gmbh/kunzmann/pull/1073 does not get updated from the "production" branch (which is our default branch) for 3 days now.

the "production" branch saw at least 10 commits in these 3 days though.

github offer me the "update branch" button to get the PR up 2 date.

**Comment by chdsbd** [comment]
@clxmstaab I did some digging but I'm not sure of the cause of this issue. Kodiak updated the pull request after I manually triggered an evaluation of the pull request.

Based on the logs it seems like Kodiak has been evaluating this PR and setting that status check you show in the screenshot. The last evaluation was `2020-07-30T13:00:44.086408601Z`.

This status check means Kodiak was hitting this code, bypassing the logic to update the pull request: https://github.com/chdsbd/kodiak/blob/0a97cf2306f26df46af0dfe72732af1ab3f32da7/bot/kodiak/evaluation.py#L621-L632


This is the code to trigger an update so for some reason the pull request wasn't meeting these requirements:
https://github.com/chdsbd/kodiak/blob/0a97cf2306f26df46af0dfe72732af1ab3f32da7/bot/kodiak/evaluation.py#L555-L560

**Comment by clxmstaab** [comment]
not sure what I can do to further investigate into the problem

**Comment by chdsbd** [comment]
@clxmstaab Are you still seeing this issue? I meant to try reproducing this over the weekend but other things got in the way. It's on my todos.

**Comment by clxmstaab** [comment]
> Are you still seeing this issue?

yes. e.g. on https://github.com/complex-gmbh/kunzmann/pull/1128 no updates happend for a few days, even though the production branch saw several commits in the meantime

![image](https://user-images.githubusercontent.com/47448731/89201959-8cb83980-d5b2-11ea-9db7-b7c3b4e09f70.png)

**Comment by chdsbd** [comment]
I was able to reproduce this so I should be able to debug this issue.

I setup a test repository using the config you provided and opened ~10 pull requests all with `WIP: ` in the title. I then merged one of them via Kodiak and found that none of the other pull requests were updated.

**Comment by chdsbd** [comment]
When the target branch is updated all of the pull requests against that are reevaluated. But the pull requests aren't being updated because the `mergeStateStatus` is `UNKNOWN`, which means the code that indicates that a pull request should be updated is false:

https://github.com/chdsbd/kodiak/blob/e80b9d125e179bc2e76e09f5e552f14a74e5aba2/bot/kodiak/evaluation.py#L546-L549

Usually when we encounter this `UNKNOWN` state we'll trigger GitHub to check the mergeability:

https://github.com/chdsbd/kodiak/blob/e80b9d125e179bc2e76e09f5e552f14a74e5aba2/bot/kodiak/evaluation.py#L675-L682



However, this code isn't being hit because the `merge.blocking_title_regex` code is hit first: 

https://github.com/chdsbd/kodiak/blob/e80b9d125e179bc2e76e09f5e552f14a74e5aba2/bot/kodiak/evaluation.py#L623-L632


So I think the solution is to always trigger a test commit when we encounter this `UNKNOWN` state. Basically we need to move some code up earlier in the `mergeability` function.


This should be quick to fix, but if you need a workaround you could use [WIP App](https://github.com/marketplace/wip) instead of the `merge.blocking_title_regex` because status checks are examined after we check for `UNKNOWN` merge states.

https://github.com/chdsbd/kodiak/blob/e80b9d125e179bc2e76e09f5e552f14a74e5aba2/bot/kodiak/evaluation.py#L741-L748

**Comment by chdsbd** [comment]
I've deployed #482 and #483, which seems to have resolved the issue. Let me know if you run into this issue again.
